kube-runtime/src/utils/stream_backoff.rs

```diff
@@ -0,0 +1,189 @@
+use std::{pin::Pin, task::Poll};
+
+use backoff::backoff::Backoff;
+use futures::{Future, Stream, TryStream};
+use pin_project::pin_project;
+use tokio::time::{sleep, Instant, Sleep};
+
+/// Applies a [`Backoff`] policy to a [`Stream`]
+///
+/// After any [`Err`] is emitted, the stream is paused for [`Backoff::next_backoff`]. The
+/// [`Backoff`] is [`reset`](`Backoff::reset`) on any [`Ok`] value.
+///
+/// If [`Backoff::next_backoff`] returns [`None`] then the backing stream is given up on, and closed.
+#[pin_project]
+pub struct StreamBackoff<S, B> {
+    #[pin]
+    stream: S,
+    backoff: B,
+    #[pin]
+    state: State,
+}
+
+#[pin_project(project = StreamBackoffStateProj)]
+// It's expected to have relatively few but long-lived `StreamBackoff`s in a project, so we would rather have
+// cheaper sleeps than a smaller `StreamBackoff`.
+#[allow(clippy::large_enum_variant)]
+enum State {
+    BackingOff(#[pin] Sleep),
+    GivenUp,
+    Awake,
+}
+
+impl<S: TryStream, B: Backoff> StreamBackoff<S, B> {
+    pub fn new(stream: S, backoff: B) -> Self {
+        Self {
+            stream,
+            backoff,
+            state: State::Awake,
+        }
+    }
+}
```

Comment:
One thought on `StreamBackoff`: because this is effectively a stream analogue of https://docs.rs/backoff/0.4.0/backoff/fn.retry.html , maybe we should allow bailing at bad errors in a simiar way to how backoff.retry defines its error type: https://docs.rs/backoff/0.4.0/backoff/enum.Error.html ?

Reply:
Hm, maybe. I'd lean towards KISS for now, and letting @ihrwein decide the future plan in the upstreaming PR (https://github.com/ihrwein/backoff/pull/50).